### PR TITLE
new component snackbar

### PR DESCRIPTION
**Pull request by yestaro**
Hi, I created a snackbar component, and add to the example.
see it on [my github page](https://yestaro.github.io/), at last navigation item

Also change FormValidator to use the snackbar when invalid

### CODE REVIEW

File: frameworks/projects/Jewel/src/main/royale/org/apache/royale/jewel/Snackbar.as [hunk 1]
```diff
@@ -0,0 +1,223 @@
...
+		{
+			if (_emphasis != value)
+            {
+                if(_emphasis)
+                {
+                    classSelectorList.toggle(_emphasis, false);
+                }
+                _emphasis = value;
+
+                classSelectorList.toggle(_emphasis, value);
```
[the remark this answers is not in the document]
Comment (yestaro) [reply frameworks/projects/Jewel/src/main/royale/org/apache/royale/jewel/Snackbar.as]: I test it in IE 11 ok. the code is the same as Jewel Button

File: frameworks/projects/Jewel/src/main/royale/org/apache/royale/jewel/beads/controllers/SnackbarController.as [hunk 2]
```diff
@@ -0,0 +1,149 @@
...
+	     *  @productversion Royale 0.9.4
+		 */
+		public function SnackbarController()
+		{
+		}
+		
+        /**
+		 * Ensure only one snackbar is created.
+		 */
+		private static var _singletonInstance:Snackbar;
```
[the remark this answers is not in the document]
Comment (yestaro) [reply frameworks/projects/Jewel/src/main/royale/org/apache/royale/jewel/beads/controllers/SnackbarController.as]: I read [Material Design Usage] (https://material.io/design/components/snackbars.html#usage) , 
Frequency
Only one snackbar may be displayed at a time.

File: frameworks/projects/Jewel/src/main/royale/org/apache/royale/jewel/beads/views/SnackbarView.as [hunk 3]
```diff
@@ -0,0 +1,135 @@
...
+        /**
+		 * The Snackbar component
+		 */
+        protected var host:Snackbar;
+
+        /**
+		 * The action element that parents the action
+		 */
+        COMPILE::JS
+		protected var actionEl:Element;
```
[the remark this answers is not in the document]
Comment (yestaro) [reply frameworks/projects/Jewel/src/main/royale/org/apache/royale/jewel/beads/views/SnackbarView.as]: OK, I'll rename it

File: frameworks/projects/Jewel/src/main/royale/org/apache/royale/jewel/beads/models/SnackbarModel.as [hunk 4]
```diff
@@ -0,0 +1,143 @@
...
+         *  
+         *  @langversion 3.0
+         *  @playerversion Flash 10.2
+         *  @playerversion AIR 2.6
+         *  @productversion Royale 0.9.4
+         */
+		public function set message(value:String):void
+		{
+            if (value == null)
+                value = "";
```
[the remark this answers is not in the document]
Comment (yestaro) [reply frameworks/projects/Jewel/src/main/royale/org/apache/royale/jewel/beads/models/SnackbarModel.as]: I didn't think so much, I think snackbar-message is the same as TextModel, so I copy the code from jewel beads TextMode.set text (https://github.com/apache/royale-asjs/blob/develop/frameworks/projects/Jewel/src/main/royale/org/apache/royale/jewel/beads/models/TextModel.as)
if undefined, the UI should display "undefined", I'll try it later

File: frameworks/projects/Jewel/src/main/royale/org/apache/royale/jewel/Snackbar.as [hunk 5]
```diff
@@ -0,0 +1,223 @@
...
+          *  @param Object parent The object that hosts the pop-up.
+          *
+          *  @langversion 3.0
+          *  @playerversion Flash 10.2
+          *  @playerversion AIR 2.6
+          *  @productversion Royale 0.9.4
+          */
+        static public function show(message:String, duration:int = 4000, actionText:String = null, parent:Object = null) : Snackbar
+		{
+            var snackbar:Snackbar = new Snackbar();
```
[the remark this answers is not in the document]
Comment (yestaro) [reply frameworks/projects/Jewel/src/main/royale/org/apache/royale/jewel/Snackbar.as]: Hi, it is 'static' function, or I misunderstood what you mean.

**Comment by carlosrovira** [comment]
HI @yestaro , thanks for the contribution. Please let us know what you plan to work, since we were very close to overlap, since I was yesterday planning on making this component! ;). So my recommendation is that you post on the list that you plan to do some task, so others could be aware and avoid working on the same.

@piotrzarzycki21, if you are ok, I can review the component and make the adaptations that could need.
Maybe is too late to get it on the release, I'll notify if I get it integrated, and if release process is not started, I can add it, if not it will be for the next! :)

thanks

**Comment by piotrzarzycki21** [comment]
@carlosrovira Actually I prefer that @yestaro do those changes with your help step by step if anything needed. Describe what is needed etc. 

 I will start RC1 hopefully tomorrow morning. You have more info on Dev list. Anything new should go on the branch.

**Comment by carlosrovira** [comment]
Hi @yestaro, 

I integrated your changes to "feature/jewel-snackbar" so we can check together the actual problems. 

1. First commit is your code with conflicts solved.
2. Second commit fix CSS rules for SASS, since I think you are not compiling SASS.  Check the comments please. There's more to know about CSS in Jewel, but I have to write a GitHub page since I still didn't share that info. I'll do as I have some time. For now CSS is ok for what we want now.

From here, you see Jewel library is failing to compile. For example, the first thing I see is that class Event is not imported in SnackBarView.as. Then it complains about "Element" Type is not found (Maybe Element is HTMLElement).

I'll stop in this point so you can check out the branch and try to compile and then we can go from there.

Thanks!

File: frameworks/projects/Jewel/src/main/royale/org/apache/royale/jewel/Snackbar.as (hunk 5 above)
[the remark this answers is not in the document]
Comment (piotrzarzycki21) [reply frameworks/projects/Jewel/src/main/royale/org/apache/royale/jewel/Snackbar.as]: Ahh...I see now :) I did miss that this method is static.

**Comment by carlosrovira** [comment]
Hi @yestaro,

I fix the compilation problems, and I found that you didn't have the class Event imported in a file and some COMPILE::JS declaration blocks in other class. That was causing the compilation fail. With that all compiles ok and work.

Can you confirm that the the code was compiling for you without the Event import and COMPILE::JS declaration blocks? if so, maybe is cause you're using a js-only setup? (I didn't try that kind of setup)

File: frameworks/projects/Jewel/src/main/royale/org/apache/royale/jewel/Snackbar.as (hunk 5 above)
[the remark this answers is not in the document]
Comment (carlosrovira) [reply frameworks/projects/Jewel/src/main/royale/org/apache/royale/jewel/Snackbar.as]: Right, this is the same the Alert component works

File: frameworks/projects/Jewel/src/main/royale/org/apache/royale/jewel/beads/controllers/SnackbarController.as (hunk 2 above)
[the remark this answers is not in the document]
Comment (carlosrovira) [reply frameworks/projects/Jewel/src/main/royale/org/apache/royale/jewel/beads/controllers/SnackbarController.as]: Hi @yestaro, if you check the demo here: https://material-components.github.io/material-components-web-catalog/#/component/snackbar, you'll see that Material web components implementation allows more that one SnackBar show at a time, although the material  document spec said the opposite...
Maybe is that they still not implemented it to allow only one.
I can position clearly, so for me is ok for now. I think we'll see the best as we get some experience with the component.

**Comment by yestaro** [comment]
Hi @carlosrovira 
I use vs code and live sass complier to develop snackbar, the asconfig.json compilerOptions.targets only for JSRoyale. 
obviously it is not enough, I will setup the full environment with maven later to confirm that for learning.
 
Next, I plan to study the reflection for Validation Metadata you mentioned, what is the first step I need to know?

**Comment by carlosrovira** [comment]
Hi @yestaro perfect. I think it's better to have all the setup.

For Validation Metadata, I must to admit that I still didn't see how to do it. Alex said that we should have all pieces to do this work already. To keep Metadata info, you should see how it's donde in other parts of the framework (Bindable, RemoteClass,...) and take a look for inspiration to GDS (Granite Data Services). If you make progress on that front I'll check that work since we need as well other libs like something similar to Swiz :). 

Thanks for working on this!! :)

**Comment by carlosrovira** [comment]
@yestaro I integrated all your code in develop now. Please take a look to each commit to see the changes I had to do, so you can continue approximating to merge with no changes. I'll make some doc about styles in Jewel in 2-3 days as I get free of some tasks I need to do, so you can be more free doing Jewel development in the future.

Congratulation once again and Thanks! :)

**Pull request closed without merging**